### PR TITLE
#161342845 Create endpoints to follow and unfollow a user

### PR DESCRIPTION
**Pull request by PriscillaSam**
#### What does this PR do?
Create endpoints to enable user follow and unfollow another user.
- follow user: **POST** `api/v1/profiles/:username/follow`.
- unfollow user: **DELETE** `api/v1/profiles/:username/follow`.

#### Description of task to be completed 
- add a `Follower` model.
- create functions to create and remove followers.
- create endpoints to follow and unfollow a user
- add a custom middleware that validates url parameters.

#### How should this be manually tested?
- on the terminal run `git clone https://github.com/andela/haven-ah-backend.git`.
- run git checkout `ft-follow-user-161342845`
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.
To test on postman

- run `npm run start-dev` and open up postman.
- to test the routes above, the user needs to be authenticated.
- signup (`localhost:5000/api/v1/users/signup`) and use email and password to log in (`localhost:5000/api/v1/users/signin`). Copy token and pass to the `x-access-token` header to test the endpoints above.

Sample responses

if following is successful, Response body shape is
```
{
    "status": 201,
    "message": "You have followed this user."
}
```
if user tries to follow his/her self
```
{
    "status": 400,
    "message": "Sorry, you can not follow yourself. You can only follow other users."
}
```
if user is already a follower
```
{
    "status": 400,
    "message": "Sorry. You already follow this user"
}
```

For **DELETE** `api/v1/profiles/:username/follow`, sample responses are
if unfollowing is successful
```
{
    "status": 200,
    "message": "You have unfollowed this user."
}
```

if a user tries to unfollow a user they don't currently follow
```
{
    "status": 400,
    "message": "Sorry. You currently don't follow this user."
}
```
if a user tries to unfollow his/her self
```
{
    "status": 400,
    "message": "Sorry, you can not unfollow yourself."
}
```

#### Any background context you want to provide?
Modifications have been made to the getUser repository functions.

#### What are the relevant pivotal tracker stories?
#161342845

#### Screenshots (if appropriate)
screen shot 2018-11-05 at 5 42 16 am

screen shot 2018-11-05 at 5 42 55 am

screen shot 2018-11-05 at 5 43 50 am

screen shot 2018-11-05 at 5 50 58 am

screen shot 2018-11-05 at 5 41 28 am

### CODE REVIEW

**Comment by victor-nweze-deel** [comment]
Looks good. Nice refactoring.

**Review by TheoOkafor — COMMENTED** [review]
Nice work Priscilla.

**Comment by ucheg6** [comment]
Good work with the refactoring... It's looking great.

File: server/repository/userRepository.js [hunk 1]
```diff
@@ -50,29 +68,44 @@ class UserRepository {
...
- * @returns {object | null} User object or null if user is not found
- */
+* Finds a user by parameter
+* @param {string} param Parameter to search by
+* @param {string} value Parameter value
+* @returns {object | null} User object or null if user is not found
+*/
+  static async getUserByParam(param, value) {
+    if (param === 'email') {
+      attributes.push('password');
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js:78]: why?

File: server/repository/userRepository.js [hunk 2]
```diff
@@ -50,29 +68,44 @@ class UserRepository {
...
+    }
+
+    const user = await User.findOne({
+      where: { [param]: value },
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js:82]: 🔥

File: swagger.js [hunk 3]
```diff
@@ -328,8 +328,7 @@ export default {
         description: 'Returns an authentication token on success.',
```
Comment (seyi-adeleke) [inline swagger.js:328]: you need to add documentation for these endpoints

**Review by seyi-adeleke — APPROVED** [review]

File: server/repository/userRepository.js (hunk 1 above)
Comment (PriscillaSam) [reply server/repository/userRepository.js:78]: @seyi-adeleke the login route needs to check that passwords match.

File: swagger.js (hunk 3 above)
Comment (PriscillaSam) [reply swagger.js:328]: @seyi-adeleke Yes, I am working on that.

File: .coveralls.yml [hunk 4]
```diff
@@ -1,2 +1,3 @@
+ service_name: travis-pro
```
Comment (seyi-adeleke) [inline .coveralls.yml]: Why? Are we paying for travis now? 😱

**Review by seyi-adeleke — APPROVED** [review]

**Pull request merged**